clang/lib/CIR/CodeGen/CIRGenModule.cpp

```diff
@@ -3187,3 +3195,68 @@ LangAS CIRGenModule::getGlobalVarAddressSpace(const VarDecl *D) {
 
   return getTargetCIRGenInfo().getGlobalVarAddressSpace(*this, D);
 }
+
+mlir::ArrayAttr CIRGenModule::buildAnnotationArgs(AnnotateAttr *attr) {
+  ArrayRef<Expr *> exprs = {attr->args_begin(), attr->args_size()};
+  if (exprs.empty()) {
+    return mlir::ArrayAttr::get(builder.getContext(), {});
+  }
+  llvm::FoldingSetNodeID id;
+  for (Expr *e : exprs) {
+    id.Add(cast<clang::ConstantExpr>(e)->getAPValueResult());
+  }
+  mlir::ArrayAttr &lookup = annotationArgs[id.ComputeHash()];
+  if (lookup)
+    return lookup;
+
+  llvm::SmallVector<mlir::Attribute, 4> args;
+  args.reserve(exprs.size());
+  for (Expr *e : exprs) {
+    if (auto *const strE =
+            ::clang::dyn_cast<clang::StringLiteral>(e->IgnoreParenCasts())) {
+      // Add trailing null character as StringLiteral->getString() does not
+      args.push_back(builder.getStringAttr(strE->getString()));
+    } else if (auto *const intE = ::clang::dyn_cast<clang::IntegerLiteral>(
+                   e->IgnoreParenCasts())) {
+      args.push_back(mlir::IntegerAttr::get(
+          mlir::IntegerType::get(builder.getContext(),
+                                 intE->getValue().getBitWidth()),
+          intE->getValue()));
+    } else {
+      llvm_unreachable("NYI");
```

Comment:
It crashes on my use-case when `e` is a `ConstantExpr` which is not handled. I can handle it with a PR but I need to understand the neighborhood first.

Reply:
OK. Actually everything is a `ConstantExpr` here. Just that I have an `enum` which is not lowered to its underlying type.
Thinking about it...

Reply:
The "normal" clang is able to evaluate the `ConstantExp` beyond just literals.
I have made https://github.com/llvm/clangir/pull/874 for integers.